### PR TITLE
Initial

### PR DESCRIPTION
**Pull request by austenstone**
Disable GHAS if repo is not in a list. Optionally enable if the repo is in the list.

Includes throttling for large number of api calls.

### CODE REVIEW

File: CODEOWNERS [hunk 1]
```diff
@@ -0,0 +1 @@
+* @actions/actions-runtime
```
Comment (NickLiffen) [inline CODEOWNERS]: What's this for 👀 This is new to me, also, there seems to be a problem, I am seeing this:

```
This CODEOWNERS file contains errors …
```

File: src/run.ts [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+      organization: {
+        repositories: {
+          nodes: repositories,
+          pageInfo: {
+            hasNextPage,
+            endCursor
+          }
+        }
+      }
+    } = await octokit.graphql(`{
```
Comment (NickLiffen) [inline src/run.ts]: I think you can be a bit more strict here with your typing. The `octokit.graphql` will return type `any`, which isn't great. For your graphql calls, I would be clearer with your typing. 

You don't have to map out the entire response, but I would expect:

```
      organization: {
        repositories: {
          nodes: repositories,
          pageInfo: {
            hasNextPage,
            endCursor
          }
        }
      }
```

To be typed, especially `repositories` as you are using that later in your `.map()` and `.filter()`

File: CODEOWNERS (hunk 1 above)
Comment (austenstone) [reply CODEOWNERS]: No need for this file. Came from a template. I'll remove it.

File: src/run.ts [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+      },
+    }
+  });
+}
+
+const run = async (): Promise<void> => {
+  try {
+    const input = getInputs();
+
+    const reposAllowed = {};
```
Comment (NickLiffen) [inline src/run.ts]: Similar to my above comment, try and be a bit more strict with your typing here 🙇 what data do you expect to be in this object? Maybe create a custom type here. Similar to how you have done:

```
  const result = {} as Input;
```

File: src/run.ts [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    } = load(readFileSync(input.file));
+    for (const [,repos] of Object.entries(teams)) {
+      repos.forEach((repo) => reposAllowed[repo] = true);
+    }
+
+    const octokit = createOctokit(input.token);
+    
+    const repoNames = await core.group('Get Repo Names', () => getRepoNames(octokit, input.org)
+      .then((repoNames) => {
+        core.setOutput('repos', JSON.stringify(repoNames));
```
Comment (NickLiffen) [inline src/run.ts]: Isn't this setting **every** repository in the organisation 🤔 versus the ones that have been changed? This get's all the repo names and outputs them all? I would expect here to maybe only show the repos that have been affected by the change? E.G

```
[{ "repoName": "x", priorState: "enabled", "currentState": "disabled"}]
```

Doing some sort of `stringify()` on only the changed repos. If they want to see the output of all the repos, that should maybe be in the log under `core.debug()`?

**Comment by NickLiffen** [comment]
@austenstone REALLY nice work 🙇 Just some small changes mostly to the typing 🙇 Excellent work and thank you for doing this 👍 💯 

Keen to get your feedback on the above 👍 

My only other thoughts, and this is more of a personal design decision, so **this is not requested changes**, but more just sharing a style idea. 

Right now the code is pretty simple and small for this, so having it all in one file makes the most sense. As this expands, and maybe we get new use cases, it may make sense to start to thinking about having a `utils` directory, and just having multiple `utils` like:

- `getInputs.ts`
- `octokit.ts`
- `getRepos.ts`
- `enableDisableGHAS.ts`

Then your `main.ts` is just a very small, simple index file that acts as a proxy which calls each util. This makes the code a bit more maintainable going forward. 

Again, for this size work, it's fine, but as this grows (if this grows), it's something to think about 🙇 

Really nice work 🌷

File: src/run.ts (hunk 2 above)
Comment (austenstone) [reply src/run.ts]: I added the type `import { Organization } from "@octokit/graphql-schema";`

File: src/run.ts (hunk 3 above)
Comment (austenstone) [reply src/run.ts]: The function `getInputs()` returns type `Input`. Seems that the type is inferred by the function call and I don't need to specify it.
Screen Shot 2022-05-31 at 2 39 19 PM

Comment (NickLiffen) [reply src/run.ts]: Sorry @austenstone, this suggested changes didn't do my any favours 😢 

I actually meant `const reposAllowed = {};`, I used `  const result = {} as Input;` as a good example. I think `const reposAllowed = {};` could use with stricter typing 🙇

Comment (austenstone) [reply src/run.ts]: I never do this but I think it's a good idea. TIL. Done!

File: src/run.ts (hunk 4 above)
Comment (austenstone) [reply src/run.ts]: output the changed repos only.

**Pull request merged**